File: repl/manager.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package repl

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"io"
	"log/slog"
	"net"
	"runtime/debug"
	"sync"
	"time"

	"turnstone/database"
	"turnstone/protocol"
)

type Source struct {
	LocalDB  string `json:"local_db"`
	RemoteDB string `json:"remote_db"`
}

type Manager struct {
	mu         sync.Mutex
	serverID   string
	peers      map[string][]Source // Addr -> List of DBs
	cancelFunc map[string]context.CancelFunc
	stores     map[string]*database.Database
	tlsConf    *tls.Config
	logger     *slog.Logger
}

// NewManager creates a manager for outgoing replication connections.
// Persistence is intentionally disabled; replication must be configured at runtime.
func NewManager(serverID string, stores map[string]*database.Database, tlsConf *tls.Config, logger *slog.Logger) *Manager {
	return &Manager{
		serverID:   serverID,
		peers:      make(map[string][]Source),
		cancelFunc: make(map[string]context.CancelFunc),
		stores:     stores,
		tlsConf:    tlsConf,
		logger:     logger,
	}
}

// IsFollowing checks if the specific database is currently configured to replicate
// from an upstream source. Used to prevent cascading repl.
func (rm *Manager) IsFollowing(dbName string) bool {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	for _, dbs := range rm.peers {
		for _, db := range dbs {
			if db.LocalDB == dbName {
				return true
			}
		}
	}
	return false
}

// Source returns the upstream address and remote database name for a given local database.
// Returns empty strings if the database is not replicating.
func (rm *Manager) Source(dbName string) (string, string) {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	for addr, sources := range rm.peers {
		for _, src := range sources {
			if src.LocalDB == dbName {
				return addr, src.RemoteDB
			}
		}
	}
	return "", ""
}

func (rm *Manager) Start() {
	rm.mu.Lock()
	defer rm.mu.Unlock()
	for addr := range rm.peers {
		rm.spawnConnection(addr)
	}
}

// removeDBFromOtherPeersLocked removes any peer entry for dbName that is
// NOT under keepAddr. A given local database must never replicate from more
// than one upstream source concurrently: if the caller bypassed the normal
// REPLICAOF/PROMOTE state-machine guard (or a stale entry was otherwise left
// behind), this ensures Follow itself stays globally deduped by dbName
// rather than only deduping within the same sourceAddr's entry. Must be
// called with rm.mu held.
func (rm *Manager) removeDBFromOtherPeersLocked(dbName, keepAddr string) {
	for addr, existingDBs := range rm.peers {
		if addr == keepAddr {
			continue
		}
		for i, db := range existingDBs {
			if db.LocalDB != dbName {
				continue
			}
			rm.logger.Warn("Replacing stale/duplicate replication source for db", "db", dbName, "old_source", addr, "new_source", keepAddr)
			newDBs := make([]Source, 0, len(existingDBs)-1)
			newDBs = append(newDBs, existingDBs[:i]...)
			newDBs = append(newDBs, existingDBs[i+1:]...)
			if len(newDBs) > 0 {
				rm.peers[addr] = newDBs
			} else {
				delete(rm.peers, addr)
				if cancel, exists := rm.cancelFunc[addr]; exists {
					cancel()
					delete(rm.cancelFunc, addr)
				}
			}
			break
		}
	}
}

func (rm *Manager) Follow(dbName, sourceAddr, sourceDB string) error {
	rm.mu.Lock()

	// Check existing
	dbs := rm.peers[sourceAddr]
	for _, db := range dbs {
		if db.LocalDB == dbName {
			rm.mu.Unlock()
			return nil // Already added
		}
	}

	// Defensively dedupe globally by dbName before we even attempt the
	// handshake, so a stale entry under a different address can't leak
	// into candidateDBs sent to the (possibly new) upstream.
	rm.removeDBFromOtherPeersLocked(dbName, sourceAddr)
	dbs = rm.peers[sourceAddr]

	// Create candidate configuration
	candidateDBs := make([]Source, len(dbs), len(dbs)+1)
	copy(candidateDBs, dbs)
	candidateDBs = append(candidateDBs, Source{LocalDB: dbName, RemoteDB: sourceDB})

	// Release lock BEFORE network call to prevent deadlock
	rm.mu.Unlock()

	// Verify Handshake Synchronously (Network Call)
	// Handshake timeout is 10s.
	if err := rm.verifyHandshake(sourceAddr, candidateDBs); err != nil {
		return err
	}

	// Re-acquire lock to apply changes
	rm.mu.Lock()
	defer rm.mu.Unlock()

	// Re-fetch current state in case it changed while we were verifying
	currentDBs := rm.peers[sourceAddr]
	for _, db := range currentDBs {
		if db.LocalDB == dbName {
			return nil // Someone else added it
		}
	}

	// Re-dedupe: another Follow for the same dbName under a different
	// address may have raced in while we were doing the network handshake.
	rm.removeDBFromOtherPeersLocked(dbName, sourceAddr)
	currentDBs = rm.peers[sourceAddr]

	// Append to the *current* authoritative list
	finalDBs := append(currentDBs, Source{LocalDB: dbName, RemoteDB: sourceDB})
	rm.peers[sourceAddr] = finalDBs

	if cancel, exists := rm.cancelFunc[sourceAddr]; exists {
		cancel()
	}
	rm.spawnConnection(sourceAddr)
	rm.logger.Info("Added replica source", "db", dbName, "source", sourceAddr, "remote_db", sourceDB)
	return nil
}

// verifyHandshake connects to the remote, sends Hello, and checks the response status.
func (rm *Manager) verifyHandshake(addr string, dbs []Source) error {
	// Increased timeout to 10s to prevent flaky "i/o timeout" errors under load
	dialer := net.Dialer{Timeout: 10 * time.Second}
	conn, err := tls.DialWithDialer(&dialer, "tcp", addr, rm.tlsConf)
	if err != nil {
		return err
	}
	defer conn.Close()

	if err := conn.SetDeadline(time.Now().Add(10 * time.Second)); err != nil {
		return err
	}

	// Construct Hello Payload
	buf := new(bytes.Buffer)
	binary.Write(buf, binary.BigEndian, uint32(1)) // Version
	binary.Write(buf, binary.BigEndian, uint32(len(rm.serverID)))
	buf.WriteString(rm.serverID)
	binary.Write(buf, binary.BigEndian, uint32(len(dbs)))

	for _, cfg := range dbs {
		var offset uint64
		if st, ok := rm.stores[cfg.LocalDB]; ok {
			offset = st.LastLogOffset()
		}
		binary.Write(buf, binary.BigEndian, uint32(len(cfg.RemoteDB)))
		buf.WriteString(cfg.RemoteDB)
		binary.Write(buf, binary.BigEndian, offset)
	}

	// Send Header
	header := make([]byte, 5)
	header[0] = protocol.OpCodeReplHello
	binary.BigEndian.PutUint32(header[1:], uint32(buf.Len()))

	if _, err := conn.Write(header); err != nil {
		return err
	}
	if _, err := conn.Write(buf.Bytes()); err != nil {
		return err
	}

	// Read Response Header
	respHead := make([]byte, 5)
	if _, err := io.ReadFull(conn, respHead); err != nil {
		return err
	}

	opCode := respHead[0]
	length := binary.BigEndian.Uint32(respHead[1:])

	// If Error, read body and return it
	if opCode == protocol.ResStatusErr {
		payload := make([]byte, length)
		if _, err := io.ReadFull(conn, payload); err != nil {
			return err
		}
		return fmt.Errorf("upstream rejected handshake: %s", string(payload))
	}

	return nil
}

// StopAll cancels every outbound replication connection.
func (rm *Manager) StopAll() {
	rm.mu.Lock()
	defer rm.mu.Unlock()
	for addr, cancel := range rm.cancelFunc {
		cancel()
		delete(rm.cancelFunc, addr)
		delete(rm.peers, addr)
		rm.logger.Debug("Stopped replication connection", "peer_addr", addr)
	}
}

func (rm *Manager) StopFollowing(dbName string) {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	for addr, dbs := range rm.peers {
		newDBs := make([]Source, 0, len(dbs))
		changed := false
		for _, db := range dbs {
			if db.LocalDB != dbName {
				newDBs = append(newDBs, db)
			} else {
				changed = true
			}
		}
		if changed {
			rm.peers[addr] = newDBs
			rm.logger.Info("Stopped replication source", "db", dbName, "source", addr)
			if cancel, exists := rm.cancelFunc[addr]; exists {
				cancel()
			}
			if len(newDBs) > 0 {
				rm.spawnConnection(addr)
			} else {
				delete(rm.peers, addr)
				delete(rm.cancelFunc, addr)
			}
			return
		}
	}
}

// safeGo runs fn in a new goroutine with panic recovery. A replication
// connection can be driven by data from a remote peer; if a parsing bug (or
// a malicious/corrupted peer) ever triggers a panic despite the bounds
// checks above, this keeps that failure scoped to the single replication
// goroutine instead of taking down the entire server process, which would
// otherwise happen since Go does not recover panics across goroutine
// boundaries.
func (rm *Manager) safeGo(name string, fn func()) {
	go func() {
		defer func() {
			if r := recover(); r != nil {
				rm.logger.Error("replication goroutine panicked", "goroutine", name, "panic", r, "stack", string(debug.Stack()))
			}
		}()
		fn()
	}()
}

func (rm *Manager) spawnConnection(addr string) {
	ctx, cancel := context.WithCancel(context.Background())
	rm.cancelFunc[addr] = cancel
	rm.safeGo("maintainConnection:"+addr, func() { rm.maintainConnection(ctx, addr) })
}

func (rm *Manager) maintainConnection(ctx context.Context, addr string) {
	peerLogger := rm.logger.With("peer_addr", addr)

	for {
		if ctx.Err() != nil {
			return
		}
		rm.mu.Lock()
		if _, ok := rm.peers[addr]; !ok {
			rm.mu.Unlock()
			return
		}
		dbs := make([]Source, len(rm.peers[addr]))
		copy(dbs, rm.peers[addr])
		rm.mu.Unlock()

		if len(dbs) == 0 {
			return
		}

		peerLogger.Debug("Attempting replication connection")

		if err := rm.connectAndSync(ctx, addr, dbs, peerLogger); err != nil {
			if ctx.Err() != nil {
				return
			}
			// WARN: Failure to connect/sync
			peerLogger.Warn("Replication sync failed, retrying in 3s", "err", err)
			select {
			case <-ctx.Done():
				return
			case <-time.After(3 * time.Second):
			}
		}
	}
}

// readLenPrefixedString reads a [Len:4][Bytes] string from buf at cursor,
// returning the string, the cursor advanced past it, and whether the read
// was in-bounds. All wire-format parsing below must use helpers like this
// (rather than slicing directly) because the data originates from a network
// peer: a truncated/malformed/malicious packet must never be able to panic
// this goroutine, since an unrecovered panic here would crash the entire
// server process, not just this replication connection.
func readLenPrefixedString(buf []byte, cursor int) (string, int, bool) {
	if cursor+4 > len(buf) {
		return "", cursor, false
	}
	n := int(binary.BigEndian.Uint32(buf[cursor : cursor+4]))
	cursor += 4
	if n < 0 || cursor+n > len(buf) {
		return "", cursor, false
	}
	return string(buf[cursor : cursor+n]), cursor + n, true
}

// connectAndSync connects to the remote, sends Hello, and checks the response status.
func (rm *Manager) connectAndSync(ctx context.Context, addr string, dbs []Source, logger *slog.Logger) error {
	dialer := net.Dialer{Timeout: 5 * time.Second}
	conn, err := tls.DialWithDialer(&dialer, "tcp", addr, rm.tlsConf)
	if err != nil {
		return err
	}

	rm.safeGo("connCloser:"+addr, func() {
		<-ctx.Done()
		_ = conn.Close()
	})
	defer conn.Close()

	logger.Info("Connected to Leader", "db_count", len(dbs))

	remoteToLocal := make(map[string][]string)
	expectedOffset := make(map[string]uint64)

	// Handshake
	// Format: [Ver:4][IDLen:4][ID][NumDBs:4] ... [NameLen:4][Name][Offset:8]
	buf := new(bytes.Buffer)
	binary.Write(buf, binary.BigEndian, uint32(1))
	binary.Write(buf, binary.BigEndian, uint32(len(rm.serverID)))
	buf.WriteString(rm.serverID)
	binary.Write(buf, binary.BigEndian, uint32(len(dbs)))

	for _, cfg := range dbs {
		var offset uint64
		if st, ok := rm.stores[cfg.LocalDB]; ok {
			offset = st.LastLogOffset()
		}
		binary.Write(buf, binary.BigEndian, uint32(len(cfg.RemoteDB)))
		buf.WriteString(cfg.RemoteDB)
		binary.Write(buf, binary.BigEndian, offset)

		logger.Debug("Sending Hello for DB", "remote_db", cfg.RemoteDB, "local_db", cfg.LocalDB, "start_log_id", offset)

		remoteToLocal[cfg.RemoteDB] = append(remoteToLocal[cfg.RemoteDB], cfg.LocalDB)
	}

	header := make([]byte, 5)
	header[0] = protocol.OpCodeReplHello
	binary.BigEndian.PutUint32(header[1:], uint32(buf.Len()))
	if _, err := conn.Write(header); err != nil {
		return err
	}
	if _, err := conn.Write(buf.Bytes()); err != nil {
		return err
	}

	respHead := make([]byte, 5)
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		if _, err := io.ReadFull(conn, respHead); err != nil {
			return err
		}

		opCode := respHead[0]
		length := binary.BigEndian.Uint32(respHead[1:])
		payload := make([]byte, length)
		if _, err := io.ReadFull(conn, payload); err != nil {
			return err
		}

		// Handle explicit error from server during streaming
		if opCode == protocol.ResStatusErr {
			return fmt.Errorf("remote error during stream: %s", string(payload))
		}

		// --- VERIFY CRC ---
		// We expect CRC for SafePoint and LogSegment
		if opCode == protocol.OpCodeReplSafePoint ||
			opCode == protocol.OpCodeReplLogRange {

			if len(payload) < 4 {
				return fmt.Errorf("packet too short for crc")
			}
			crcReceived := binary.BigEndian.Uint32(payload[:4])
			rawBody := payload[4:]
			if crc32.Checksum(rawBody, protocol.Crc32Table) != crcReceived {
				return fmt.Errorf("crc mismatch on replication stream")
			}
			// Strip CRC for logic processing
			payload = rawBody
		}

		// --- Handle Safe Point Propagation ---
		if opCode == protocol.OpCodeReplSafePoint {
			cursor := 0
			if cursor+4 > len(payload) {
				continue
			}
			nLen := int(binary.BigEndian.Uint32(payload[cursor : cursor+4]))
			cursor += 4
			if cursor+nLen > len(payload) {
				continue
			}
			remoteDBName := string(payload[cursor : cursor+nLen])
			cursor += nLen
			cursor += 4 // Skip Count/Reserved

			if cursor+8 <= len(payload) {
				safeOffset := binary.BigEndian.Uint64(payload[cursor:])

				if localDBNames, ok := remoteToLocal[remoteDBName]; ok {
					for _, localDB := range localDBNames {
						if st, ok := rm.stores[localDB]; ok {
							st.SetLeaderRetainOffset(safeOffset)
						}
					}
				}
			}
			continue
		}

		// --- Handle raw log byte segment (physical replication) ---
		if opCode == protocol.OpCodeReplLogRange {
			remoteDBName, cursor, ok := readLenPrefixedString(payload, 0)
			if !ok || cursor+4 > len(payload) {
				return fmt.Errorf("malformed log segment packet (len=%d)", len(payload))
			}
			cursor += 4 // skip count/reserved
			if cursor+16 > len(payload) {
				return fmt.Errorf("malformed log segment header (len=%d)", len(payload))
			}
			startOff := binary.BigEndian.Uint64(payload[cursor : cursor+8])
			endOff := binary.BigEndian.Uint64(payload[cursor+8 : cursor+16])
			cursor += 16
			segData := payload[cursor:]
			if endOff < startOff || int(endOff-startOff) != len(segData) {
				return fmt.Errorf("log segment offset mismatch (start=%d end=%d len=%d)", startOff, endOff, len(segData))
			}

			if localDBNames, ok := remoteToLocal[remoteDBName]; ok {
				for _, localDBName := range localDBNames {
					if st, ok := rm.stores[localDBName]; ok {
						exp, ok := expectedOffset[localDBName]
						if !ok {
							exp = st.LastLogOffset()
						}
						if startOff != exp {
							return fmt.Errorf("log range start mismatch for db %s: got %d want %d", localDBName, startOff, exp)
						}
						if _, err := st.ApplyLogRange(segData); err != nil {
							logger.Error("Failed to apply log segment", "db", localDBName, "err", err)
							return err
						}
						expectedOffset[localDBName] = endOff
						if endOff > 0 {
							ackBuf := new(bytes.Buffer)
							binary.Write(ackBuf, binary.BigEndian, uint32(len(remoteDBName)))
							ackBuf.WriteString(remoteDBName)
							binary.Write(ackBuf, binary.BigEndian, endOff)

							h := make([]byte, 5)
							h[0] = protocol.OpCodeReplAck
							binary.BigEndian.PutUint32(h[1:], uint32(ackBuf.Len()))

							if _, err := conn.Write(h); err != nil {
								return err
							}
							if _, err := conn.Write(ackBuf.Bytes()); err != nil {
								return err
							}
						}
					}
				}
			}
			continue
		}
	}
}

File: rust/crates/turnstone-repl/src/manager.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::collections::HashMap;
use std::io::{Read, Write};
use std::net::TcpStream;
use std::sync::Arc;
use std::time::Duration;

use parking_lot::Mutex;
use rustls::pki_types::ServerName;
use turnstone_database::Database;
use turnstone_protocol::{
    decode_header, encode_frame, OP_REPL_ACK, OP_REPL_HELLO, OP_REPL_LOG_RANGE, OP_REPL_SAFE_POINT,
    RES_ERR,
};

#[derive(Debug, Clone)]
pub struct Source {
    pub local_db: String,
    pub remote_db: String,
}

pub struct Manager {
    server_id: String,
    peers: Arc<Mutex<HashMap<String, Vec<Source>>>>,
    cancel: Arc<Mutex<HashMap<String, std::sync::mpsc::Sender<()>>>>,
    stores: HashMap<String, Arc<Database>>,
    tls: Arc<rustls::ClientConfig>,
}

impl Manager {
    pub fn new(
        server_id: impl Into<String>,
        stores: HashMap<String, Arc<Database>>,
        tls: Arc<rustls::ClientConfig>,
    ) -> Self {
        Self {
            server_id: server_id.into(),
            peers: Arc::new(Mutex::new(HashMap::new())),
            cancel: Arc::new(Mutex::new(HashMap::new())),
            stores,
            tls,
        }
    }

    pub fn start(&self) {
        let addrs: Vec<String> = self.peers.lock().keys().cloned().collect();
        for addr in addrs {
            self.spawn_connection(&addr);
        }
    }

    pub fn stop_all(&self) {
        let mut cancel = self.cancel.lock();
        for (_, tx) in cancel.drain() {
            let _ = tx.send(());
        }
        self.peers.lock().clear();
    }

    pub fn is_following(&self, db_name: &str) -> bool {
        self.peers
            .lock()
            .values()
            .any(|dbs| dbs.iter().any(|db| db.local_db == db_name))
    }

    pub fn source(&self, db_name: &str) -> (String, String) {
        for (addr, sources) in self.peers.lock().iter() {
            for src in sources {
                if src.local_db == db_name {
                    return (addr.clone(), src.remote_db.clone());
                }
            }
        }
        (String::new(), String::new())
    }

    pub fn follow(
        &self,
        db_name: &str,
        source_addr: &str,
        source_db: &str,
    ) -> Result<(), String> {
        {
            let peers = self.peers.lock();
            if let Some(dbs) = peers.get(source_addr) {
                if dbs.iter().any(|db| db.local_db == db_name) {
                    return Ok(());
                }
            }
        }

        self.remove_db_from_other_peers(db_name, source_addr);

        let mut candidate = self.peers.lock().get(source_addr).cloned().unwrap_or_default();
        candidate.push(Source {
            local_db: db_name.to_string(),
            remote_db: source_db.to_string(),
        });

        self.verify_handshake(source_addr, &candidate)?;

        self.remove_db_from_other_peers(db_name, source_addr);
        {
            let mut peers = self.peers.lock();
            let mut current = peers.get(source_addr).cloned().unwrap_or_default();
            if current.iter().any(|db| db.local_db == db_name) {
                return Ok(());
            }
            current.push(Source {
                local_db: db_name.to_string(),
                remote_db: source_db.to_string(),
            });
            peers.insert(source_addr.to_string(), current);
        }

        if let Some(tx) = self.cancel.lock().remove(source_addr) {
            let _ = tx.send(());
        }
        self.spawn_connection(source_addr);
        Ok(())
    }

    pub fn stop_following(&self, db_name: &str) {
        let mut addr_to_restart: Option<String> = None;
        let mut empty_addr: Option<String> = None;
        {
            let mut peers = self.peers.lock();
            for (addr, dbs) in peers.iter_mut() {
                let before = dbs.len();
                dbs.retain(|db| db.local_db != db_name);
                if dbs.len() != before {
                    addr_to_restart = Some(addr.clone());
                    if dbs.is_empty() {
                        empty_addr = Some(addr.clone());
                    }
                    break;
                }
            }
            if let Some(addr) = empty_addr {
                peers.remove(&addr);
            }
        }
        if let Some(addr) = addr_to_restart {
            if let Some(tx) = self.cancel.lock().remove(&addr) {
                let _ = tx.send(());
            }
            if self.peers.lock().contains_key(&addr) {
                self.spawn_connection(&addr);
            }
        }
    }

    fn remove_db_from_other_peers(&self, db_name: &str, keep_addr: &str) {
        let mut cancel_addrs = Vec::new();
        {
            let mut peers = self.peers.lock();
            for (addr, dbs) in peers.iter_mut() {
                if addr == keep_addr {
                    continue;
                }
                let before = dbs.len();
                dbs.retain(|db| db.local_db != db_name);
                if dbs.len() != before {
                    if dbs.is_empty() {
                        cancel_addrs.push(addr.clone());
                    }
                }
            }
            for addr in &cancel_addrs {
                peers.remove(addr);
            }
        }
        for addr in cancel_addrs {
            if let Some(tx) = self.cancel.lock().remove(&addr) {
                let _ = tx.send(());
            }
        }
    }

    fn verify_handshake(&self, addr: &str, dbs: &[Source]) -> Result<(), String> {
        let mut stream = self.connect(addr)?;
        self.send_hello(&mut stream, dbs)?;
        let (op, payload) = read_frame(&mut stream)?;
        if op == RES_ERR {
            return Err(format!(
                "upstream rejected handshake: {}",
                String::from_utf8_lossy(&payload)
            ));
        }
        Ok(())
    }

    fn spawn_connection(&self, addr: &str) {
        let (tx, rx) = std::sync::mpsc::channel();
        self.cancel.lock().insert(addr.to_string(), tx);
        let addr = addr.to_string();
        let server_id = self.server_id.clone();
        let tls = self.tls.clone();
        let stores = self
            .stores
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect::<HashMap<_, _>>();
        let peers = Arc::clone(&self.peers);
        let cancel_map = Arc::clone(&self.cancel);
        std::thread::spawn(move || {
            let mgr = Manager {
                server_id,
                peers,
                cancel: cancel_map,
                stores,
                tls,
            };
            mgr.maintain_connection(&addr, rx);
        });
    }

    fn maintain_connection(&self, addr: &str, cancel: std::sync::mpsc::Receiver<()>) {
        let mut retry = Duration::from_millis(100);
        loop {
            if cancel.try_recv().is_ok() {
                return;
            }
            let dbs = self.peers.lock().get(addr).cloned().unwrap_or_default();
            if dbs.is_empty() {
                return;
            }
            if self.connect_and_sync(addr, &dbs, &cancel).is_err() {
                if cancel.try_recv().is_ok() {
                    return;
                }
                std::thread::sleep(retry);
                retry = (retry * 2).min(Duration::from_secs(3));
                continue;
            }
            retry = Duration::from_millis(100);
        }
    }

    fn connect_and_sync(
        &self,
        addr: &str,
        dbs: &[Source],
        cancel: &std::sync::mpsc::Receiver<()>,
    ) -> Result<(), String> {
        let mut stream = self.connect(addr)?;
        self.send_hello(&mut stream, dbs)?;

        let mut remote_to_local: HashMap<String, Vec<String>> = HashMap::new();
        let mut expected_offset: HashMap<String, u64> = HashMap::new();
        for cfg in dbs {
            remote_to_local
                .entry(cfg.remote_db.clone())
                .or_default()
                .push(cfg.local_db.clone());
        }

        loop {
            if cancel.try_recv().is_ok() {
                return Err("cancelled".into());
            }
            let (op, payload) = read_frame(&mut stream)?;
            if op == RES_ERR {
                return Err(format!(
                    "remote error during stream: {}",
                    String::from_utf8_lossy(&payload)
                ));
            }

            let payload = if op == OP_REPL_SAFE_POINT || op == OP_REPL_LOG_RANGE {
                verify_crc_payload(&payload)?
            } else {
                payload
            };

            if op == OP_REPL_SAFE_POINT {
                handle_safe_point(&payload, &remote_to_local, &self.stores);
                continue;
            }

            if op == OP_REPL_LOG_RANGE {
                apply_log_segment(
                    &payload,
                    &remote_to_local,
                    &self.stores,
                    &mut expected_offset,
                    &mut stream,
                )?;
                continue;
            }
        }
    }

    fn connect(
        &self,
        addr: &str,
    ) -> Result<rustls::StreamOwned<rustls::ClientConnection, TcpStream>, String> {
        use std::net::ToSocketAddrs;
        let sock = addr
            .to_socket_addrs()
            .map_err(|e| e.to_string())?
            .next()
            .ok_or_else(|| "resolve failed".to_string())?;
        let tcp = TcpStream::connect_timeout(&sock, Duration::from_secs(10))
            .map_err(|e| e.to_string())?;
        tcp.set_read_timeout(Some(Duration::from_secs(10)))
            .map_err(|e| e.to_string())?;
        tcp.set_write_timeout(Some(Duration::from_secs(10)))
            .map_err(|e| e.to_string())?;
        let server_name = ServerName::try_from("localhost".to_string()).map_err(|_| "bad sni")?;
        let conn = rustls::ClientConnection::new(self.tls.clone(), server_name)
            .map_err(|e| e.to_string())?;
        Ok(rustls::StreamOwned::new(conn, tcp))
    }

    fn send_hello(
        &self,
        stream: &mut rustls::StreamOwned<rustls::ClientConnection, TcpStream>,
        dbs: &[Source],
    ) -> Result<(), String> {
        let mut body = Vec::new();
        body.extend_from_slice(&1u32.to_be_bytes());
        body.extend_from_slice(&(self.server_id.len() as u32).to_be_bytes());
        body.extend_from_slice(self.server_id.as_bytes());
        body.extend_from_slice(&(dbs.len() as u32).to_be_bytes());
        for cfg in dbs {
            let offset = self
                .stores
                .get(&cfg.local_db)
                .map(|st| st.last_log_offset())
                .unwrap_or(0);
            body.extend_from_slice(&(cfg.remote_db.len() as u32).to_be_bytes());
            body.extend_from_slice(cfg.remote_db.as_bytes());
            body.extend_from_slice(&offset.to_be_bytes());
        }
        write_frame(stream, OP_REPL_HELLO, &body)
    }
}

fn write_frame(
    stream: &mut rustls::StreamOwned<rustls::ClientConnection, TcpStream>,
    op: u8,
    body: &[u8],
) -> Result<(), String> {
    let frame = encode_frame(op, body);
    stream.write_all(&frame).map_err(|e| e.to_string())?;
    stream.flush().map_err(|e| e.to_string())?;
    Ok(())
}

fn read_frame(
    stream: &mut rustls::StreamOwned<rustls::ClientConnection, TcpStream>,
) -> Result<(u8, Vec<u8>), String> {
    let mut head = [0u8; 5];
    stream.read_exact(&mut head).map_err(|e| e.to_string())?;
    let hdr = decode_header(&head).ok_or_else(|| "bad header".to_string())?;
    let mut payload = vec![0u8; hdr.payload_len as usize];
    if !payload.is_empty() {
        stream.read_exact(&mut payload).map_err(|e| e.to_string())?;
    }
    Ok((hdr.status_or_op, payload))
}

fn verify_crc_payload(payload: &[u8]) -> Result<Vec<u8>, String> {
    if payload.len() < 4 {
        return Err("packet too short for crc".into());
    }
    let crc_received = u32::from_be_bytes(payload[..4].try_into().unwrap());
    let body = &payload[4..];
    if crc32fast::hash(body) != crc_received {
        return Err("crc mismatch on replication stream".into());
    }
    Ok(body.to_vec())
}

fn read_len_prefixed_string(buf: &[u8], cursor: usize) -> Option<(String, usize)> {
    if cursor + 4 > buf.len() {
        return None;
    }
    let n = u32::from_be_bytes(buf[cursor..cursor + 4].try_into().ok()?) as usize;
    let start = cursor + 4;
    if start + n > buf.len() {
        return None;
    }
    Some((
        String::from_utf8_lossy(&buf[start..start + n]).into_owned(),
        start + n,
    ))
}

fn handle_safe_point(
    payload: &[u8],
    remote_to_local: &HashMap<String, Vec<String>>,
    stores: &HashMap<String, Arc<Database>>,
) {
    let (remote_db, mut cursor) = match read_len_prefixed_string(payload, 0) {
        Some(v) => v,
        None => return,
    };
    cursor += 4;
    if cursor + 8 > payload.len() {
        return;
    }
    let safe = u64::from_be_bytes(payload[cursor..cursor + 8].try_into().unwrap());
    if let Some(locals) = remote_to_local.get(&remote_db) {
        for local in locals {
            if let Some(st) = stores.get(local) {
                st.set_leader_retain_offset(safe);
            }
        }
    }
}

fn apply_log_segment(
    payload: &[u8],
    remote_to_local: &HashMap<String, Vec<String>>,
    stores: &HashMap<String, Arc<Database>>,
    expected_offset: &mut HashMap<String, u64>,
    stream: &mut rustls::StreamOwned<rustls::ClientConnection, TcpStream>,
) -> Result<(), String> {
    let (remote_db, mut cursor) =
        read_len_prefixed_string(payload, 0).ok_or_else(|| "malformed log segment packet".to_string())?;
    cursor += 4;
    if cursor + 16 > payload.len() {
        return Err("malformed log segment header".into());
    }
    let start_off = u64::from_be_bytes(payload[cursor..cursor + 8].try_into().unwrap());
    let end_off = u64::from_be_bytes(payload[cursor + 8..cursor + 16].try_into().unwrap());
    cursor += 16;
    let seg_data = &payload[cursor..];
    if end_off < start_off || (end_off - start_off) as usize != seg_data.len() {
        return Err(format!(
            "log segment offset mismatch (start={start_off} end={end_off} len={})",
            seg_data.len()
        ));
    }

    let locals = remote_to_local
        .get(&remote_db)
        .cloned()
        .unwrap_or_default();
    for local_db in locals {
        let Some(st) = stores.get(&local_db) else {
            continue;
        };
        let mut exp = expected_offset
            .get(&local_db)
            .copied()
            .unwrap_or_else(|| st.last_log_offset());
        if start_off != exp {
            if exp == 0 && start_off > 0 && st.last_log_offset() == 0 && !seg_data.is_empty() {
                st.init_log_at_lsn(start_off).map_err(|e| e.to_string())?;
                exp = start_off;
            } else {
                return Err(format!(
                    "log range start mismatch for db {local_db}: got {start_off} want {exp}"
                ));
            }
        }
        st.apply_log_range(seg_data).map_err(|e| e.to_string())?;
        expected_offset.insert(local_db.clone(), end_off);
        if end_off > 0 {
            let mut ack = Vec::new();
            ack.extend_from_slice(&(remote_db.len() as u32).to_be_bytes());
            ack.extend_from_slice(remote_db.as_bytes());
            ack.extend_from_slice(&end_off.to_be_bytes());
            write_frame(stream, OP_REPL_ACK, &ack)?;
        }
    }
    Ok(())
}
